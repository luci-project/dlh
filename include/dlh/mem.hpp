// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#pragma once

#include <dlh/types.hpp>
#include <dlh/math.hpp>

namespace Memory {
/*! \brief Allocate a memory block
 * The memory is not initialized.
 *
 * \param size Requested size of memory in bytes.
 * \return Address of memory or `0` on error (no memory available) or if size
 *         was zero.
 */
uintptr_t alloc(size_t size);

template<typename T>
inline T* alloc(size_t size) {
	return reinterpret_cast<T*>(alloc(size));
}

/*! \brief Free an allocated memory block
 *
 * \param addr Address of an previously allocated memory block.
 */
void free(uintptr_t addr);

template<typename T>
inline void free(T* addr) {
	return free(reinterpret_cast<uintptr_t>(addr));
}


/*! \brief Change the size of an allocated memory block
 * The contents will be unchanged in the range from the start of the region up
 * to the minimum of the old and new sizes. If the new size is larger than the
 * old size, the added memory will not be initialized.
 *
 * \param addr Address of an previously allocated memory block.
 *             If `0`, then the call is equivalent to \ref alloc().
 * \param size New size of the memory block. If equal to zero, then the call
 *             is equivalent to free
 * \return Address of new memory block or `0` on error
 */
uintptr_t realloc(uintptr_t addr, size_t size);

template<typename T>
inline T* realloc(T* addr, size_t size) {
	return reinterpret_cast<T*>(realloc(reinterpret_cast<uintptr_t>(addr), size));
}

/*! \brief Allocate memory for an array of elements
 * The memory is set to zero.
 *
 * \param nmemb Number of elements
 * \param size Size of an element in bytes
 * \return address of the allocated memory or `0` if the request fails
 */
uintptr_t alloc_array(size_t nmemb, size_t size);

template<typename T>
inline T* alloc_array(size_t nmemb) {
	return reinterpret_cast<T*>(alloc_array(nmemb, sizeof(T)));
}

/*! \brief Copy a memory area
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return Address of destination
 * \note The memory must not overlap!
 */
inline uintptr_t copy(uintptr_t dest, uintptr_t src, size_t size) {
	unsigned char * __restrict__ destination = reinterpret_cast<unsigned char*>(dest);
	unsigned char const * __restrict__ source = reinterpret_cast<unsigned char const*>(src);

	for(size_t i = 0; i != size; ++i) {
		destination[i] = source[i];
	}

	return dest;
}

/*! \brief Copy a memory area
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return pointer to destination
 * \note The memory must not overlap!
 */
template<typename T, typename U>
inline T* copy(T * __restrict__ dest, U const * __restrict__ src, size_t size) {
	return reinterpret_cast<T*>(copy(reinterpret_cast<uintptr_t>(dest), reinterpret_cast<uintptr_t>(src), size));
}

/*! \brief Copy a memory area
 * while the source may overlap with the destination
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return Address of destination
 */
inline uintptr_t move(uintptr_t dest, uintptr_t src, size_t size) {
	unsigned char * __restrict__ destination = reinterpret_cast<unsigned char*>(dest);
	unsigned char const * __restrict__ source = reinterpret_cast<unsigned char const*>(src);

	if(source > destination) {
		for(size_t i = 0; i != size; ++i)
			destination[i] = source[i];
	} else {
		for(size_t i = size; i != 0; --i)
			destination[i-1] = source[i-1];
	}

	return dest;
}

/*! \brief Copy a memory area
 * while the source may overlap with the destination
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return pointer to destination
 */
template<typename T, typename U>
inline T* move(T * dest, U const * src, size_t size) {
	return reinterpret_cast<T*>(move(reinterpret_cast<uintptr_t>(dest), reinterpret_cast<uintptr_t>(src), size));
}

/*! \brief Fill a memory area with a pattern
 * \ingroup string
 * \param dest destination buffer
 * \param pattern single byte pattern
 * \param size number of bytes to fill with pattern
 * \return address of destination
 */
inline uintptr_t set(uintptr_t dest, int pattern, size_t size) {
	unsigned char *destination = reinterpret_cast<unsigned char*>(dest);

	for(size_t i = 0; i != size; ++i) {
		destination[i] = static_cast<unsigned char>(pattern);
	}

	return dest;
}

/*! \brief Fill a memory area with a pattern
 * \ingroup string
 * \param dest destination buffer
 * \param pattern single byte pattern
 * \param size number of bytes to fill with pattern
 * \return pointer to destination
 */
template<typename T>
inline T* set(T* dest, int pattern, size_t size) {
	return reinterpret_cast<T*>(set(reinterpret_cast<uintptr_t>(dest), pattern, size));
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 pointer to first element
 * \param s2 pointer to second element
 * \param size number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
inline int compare(uintptr_t s1, uintptr_t s2, size_t n) {
	const unsigned char * c1 = reinterpret_cast<const unsigned char*>(s1);
	const unsigned char * c2 = reinterpret_cast<const unsigned char*>(s2);

	for(size_t i = 0; i != n; ++i)
		if (c1[i] != c2[i])
			return static_cast<int>(c1[i]) - static_cast<int>(c2[i]);

	return 0;
}


/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 pointer to first element
 * \param s2 pointer to second element
 * \param size number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T * s1, const U * s2, size_t n) {
	return compare(reinterpret_cast<uintptr_t>(s1), reinterpret_cast<uintptr_t>(s2), n);
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 pointer to first element
 * \param s2 pointer to second element
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T * s1, const U * s2) {
	int r = compare(s1, s2, Math::min(sizeof(T), sizeof(U)));
	return r == 0 ? sizeof(U) - sizeof(T) : r;
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 reference to first element
 * \param s2 reference to second element
 * \param size number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T & s1, const T & s2, size_t n) {
	return compare(reinterpret_cast<uintptr_t>(&s1), reinterpret_cast<uintptr_t>(&s2), n);
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 reference to first element
 * \param s2 reference to second element
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T & s1, const U & s2) {
	int r = compare(reinterpret_cast<uintptr_t>(&s1), reinterpret_cast<uintptr_t>(&s2), Math::min(sizeof(T), sizeof(U)));
	return r == 0 ? sizeof(U) - sizeof(T) : r;
}

}  // namespace Memory
