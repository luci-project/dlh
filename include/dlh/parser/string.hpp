#pragma once

#include <dlh/container/vector.hpp>
#include <dlh/types.hpp>

namespace Parser {

bool string(unsigned long long & target, const char * value);

bool string(long long & target, const char * value);

bool string(unsigned long & target, const char * value);

bool string(long & target, const char * value);

bool string(unsigned int & target, const char * value);

bool string(int & target, const char * value);

bool string(unsigned short & target, const char * value);

bool string(short & target, const char * value);

bool string(unsigned char & target, const char * value);

bool string(char & target, const char * value);

bool string(bool & target, const char * value);

bool string(const char * & target, const char * value);

template<typename T>
bool string(Vector<T> & target, const char * value) {
	T tmp;
	bool r = string(tmp, value);
	target.push_back(tmp);
	return r;
}

}  // namespace Parse
