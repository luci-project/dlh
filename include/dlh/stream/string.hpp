// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#pragma once

#include <dlh/types.hpp>
#include <dlh/stream/buffer.hpp>

template<size_t BUFFERSZ>
class StringStream : public BufferStream {
 protected:
	char buffer[BUFFERSZ];

 public:
	/*! \brief Default constructor  */
	StringStream() : BufferStream(buffer, BUFFERSZ) {}

	/*! \brief Destructor */
	virtual ~StringStream() {}
};
