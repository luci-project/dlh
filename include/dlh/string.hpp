#pragma once

#include <dlh/types.hpp>
#include <dlh/container/vector.hpp>

namespace String {

/*! \brief Find the first occurrence of a character in a string
 * \param s string to search in
 * \param c character to find
 * \return Pointer to first occurrence of the character
 *         or to null byte at the end of the string if not found
 */
const char* find_or_end(const char *s, int c);

/*! \brief Find the first occurrence of a substring in a string
 * \param haystack string to search in
 * \param needle string to find
 * \return Pointer to first occurrence of the needle
 *         or to null byte at the end of the string if not found
 */
const char* find_or_end(const char *haystack, const char *needle);

/*! \brief Find the first occurrence of a character in a string
 * \param s string to search in
 * \param c character to find
 * \return Pointer to first occurrence of the character
 *         or to nullptr if not found
 */
const char* find(const char *s, int c);

/*! \brief Find the first occurrence of a substring in a string
 * \param haystack string to search in
 * \param needle string to find
 * \return Pointer to first occurrence of the needle
 *         or to nullptr if not found
 */
const char* find(const char *haystack, const char *needle);

/*! \brief Find the last occurrence of a character in a string
 * \param s string to search in
 * \param c character to find
 * \return Pointer to last occurrence of the character
 *         or to nullptr if not found
 */
const char* find_last(const char *s, int c);

/*! \brief Find the last occurrence of a substring in a string
 * \param haystack string to search in
 * \param needle string to find
 * \return Pointer to last occurrence of the needle
 *         or to nullptr if not found
 */
const char* find_last(const char *haystack, const char *needle);

/*! \brief Replace characters in a string
 * \param target pointer to string (target memory will be modified!)
 * \param from character to be replaced
 * \param to replacement character
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 */
char * replace_inplace(char *target, int from, int to, size_t max = SIZE_MAX);

/*! \brief Replace characters in a string
 * \param target pointer to string (string will be copied)
 * \param from character to be replaced
 * \param to replacement character
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 * \note Target pointer has to be freed after usage
 */
char * replace(const char *target, int from, int to, size_t max = SIZE_MAX);

/*! \brief Replace stubstrings in a string
 * \param target pointer to string  (target memory will be modified!)
 * \param from substring to be replaced
 * \param to replacement substring (must be equal or shorter than `from`)
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 */
char * replace_inplace(char *target, const char *from, const char * to, size_t max = SIZE_MAX);

/*! \brief Replace stubstrings in a string
 * \param target pointer to string (string will be copied)
 * \param from substring to be replaced
 * \param to replacement substring
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 * \note Target pointer has to be freed after usage
 */
char * replace(const char *target, const char *from, const char * to, size_t max = SIZE_MAX);

/*! \brief Compare two strings
 * \param s1 first string
 * \param s2 second string
 * \return an integer less than, equal to, or greater than zero if first string is found, respectively,
 *         to be less than, to match, or be greater than second string
 */
int compare(const char *s1, const char *s2);

/*! \brief Compare two strings
 * \param s1 first string
 * \param s2 second string
 * \param n number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first string are
 *          found, respectively, to be less than, to match, or be greater than second string
 */
int compare(const char *s1, const char *s2, size_t n);

/*! \brief Calculate the length of a string
 * \param s pointer to a string
 * \return number of bytes in the string
 */
size_t len(const char *s);

/*! \brief Copy the contents of a string
 * including the terminating null byte (`\0`)
 * \param dest destination string buffer
 * \param src source string buffer
 * \return a pointer to the destination string buffer
 * \note Beware of buffer overruns!
 */
char* copy(char * dest, const char * src);  //NOLINT

/*! \brief Copy the contents of a string up to a maximum length
 * or the terminating null byte (`\0`), whatever comes first.
 * \param dest destination string buffer
 * \param src source string buffer
 * \param n maximum number of bytes to copy
 * \return a pointer to the destination string buffer
 * \note If there is no null byte (`\0`) among the first `n` bytes, the destination will not be null-terminated!
 */
char* copy(char * dest, const char * src, size_t n);

/*! \brief Duplicate a string
 * \param s pointer to a string
 * \return pointer to a duplicated string allocated with malloc or `nullptr` if allocation failed.
 */
char* duplicate(const char *s);

/*! \brief Duplicate a string
 * \param s pointer to a string
 * \param n maximum length
 * \return pointer to a duplicated string (up to maximum length, always null-terminated),
 *         allocated with malloc or `nullptr` if allocation failed.
 */
char* duplicate(const char *s, size_t n);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string (target memory will be modified!)
 * \param delimiter character to split string (will be overwritten with NULL character)
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 */
Vector<const char *> split_inplace(char * source, int delimiter, size_t max = SIZE_MAX);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string
 * \param delimiter character to split string
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 * \note You have to free each element of the result vector!
 */
Vector<const char *> split(const char * source, int delimiter, size_t max = SIZE_MAX);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string (target memory will be modified!)
 * \param delimiter substring to split string
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 */
Vector<const char *> split_inplace(char * source, const char * delimiter, size_t max = SIZE_MAX);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string
 * \param delimiter substring to split string
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 * \note You have to free each element of the result vector!
 */
Vector<const char *> split(const char * source, const char * delimiter, size_t max = SIZE_MAX);
}  // namespace String
