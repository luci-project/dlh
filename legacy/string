#pragma once

#include <dlh/strptr.hpp>

namespace std {

using string = StrPtr;

} // namespace std
