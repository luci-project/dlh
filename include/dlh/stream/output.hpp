#pragma once

#include <dlh/syscall.hpp>
#include <dlh/stream/string.hpp>

template<size_t BUFFERSZ>
class OutputStream : public StringStream<BUFFERSZ> {
 protected:
	int fd;

 public:
	/*! \brief Default constructor  */
	explicit OutputStream(int fd) : fd(fd) {}

	/*! \brief Destructor */
	virtual ~OutputStream() {}

	/*! \brief Flush the buffer.
	 */
	virtual void flush() override {
		size_t l = 0;
		while (l < this->_pos) {
			auto r = Syscall::write(fd, this->_bufptr + l, this->_pos - l);
			if (!r.valid())
				break;
			else
				l += r.value();
		}
		this->_pos = 0;
	}

};

extern OutputStream<1024> cout, cerr;
