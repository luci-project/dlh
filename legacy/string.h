#pragma once

#include <dlh/string.hpp>
