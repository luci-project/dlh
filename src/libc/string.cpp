// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#include <dlh/string.hpp>
#include <dlh/mem.hpp>

#ifdef DLH_LEGACY

extern "C" char *strchrnul(const char *s, int c) {
	return const_cast<char*>(String::find_or_end(s, c));
}

extern "C" char *strchr(const char *s, int c) {
	return const_cast<char*>(String::find(s, c));
}

extern "C" char *strstr(const char *haystack, const char *needle) {
	return const_cast<char*>(String::find(haystack, needle));
}

extern "C" char *strrchr(const char *s, int c) {
	return const_cast<char*>(String::find_last(s, c));
}

extern "C" int strcmp(const char *s1, const char *s2) {
	return String::compare(s1, s2);
}

extern "C" int strncmp(const char *s1, const char *s2, size_t n) {
	return String::compare(s1, s2, n);
}

extern "C" size_t strlen(const char *s) {
	return String::len(s);
}

extern "C" char * strcpy(char *dest, const char *src) { //NOLINT
	return String::copy(dest, src);
}

extern "C" char * strncpy(char *dest, const char *src, size_t n) {
	return String::copy(dest, src, n, true);
}

extern "C" char * strdup(const char *s) {
	return String::duplicate(s);
}

extern "C" char * strndup(const char *s, size_t n) {
	return String::duplicate(s, n);
}
#endif

extern "C" void* memcpy(void * __restrict__ dest, void const * __restrict__ src, size_t size) {
	return Memory::copy(dest, src, size);
}

extern "C" void* memmove(void * dest, void const * src, size_t size) {
	return Memory::move(dest, src, size);
}

extern "C" void* memset(void *dest, int pattern, size_t size) {
	return Memory::set(dest, pattern, size);
}

extern "C" int memcmp(const void * s1, const void * s2, size_t n) {
	return Memory::compare(s1, s2, n);
}