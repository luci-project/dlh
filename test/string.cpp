#include <dlh/stream/output.hpp>
#include <dlh/string.hpp>
#include <dlh/mem.hpp>
#include <dlh/error.hpp>

int main(int argc, const char *argv[]) {
	(void) argc;
	(void) argv;

	const char * s = "Rhabarberbarbarabarbarbaren!";

	cout << "String::find(s, 'b') = " << String::find(s, 'b') << endl;
	cout << "String::find_or_end(s, 'b') = " << String::find_or_end(s, 'b') << endl;
	cout << "String::find_last(s, 'b') = " << String::find_last(s, 'b') << endl;

	cout << "String::find(s, \"barbar\") = " << String::find(s, "barbar") << endl;
	cout << "String::find_or_end(s, \"barbar\") = " << String::find_or_end(s, "barbar") << endl;
	cout << "String::find_last(s, \"barbar\") = " << String::find_last(s, "barbar") << endl;

	cout << "String::find(s, 'y') = " << String::find(s, 'y') << endl;
	cout << "String::find_or_end(s, 'y') = " << String::find_or_end(s, 'y') << endl;
	cout << "String::find_last(s, 'y') = " << String::find_last(s, 'y') << endl;

	cout << "String::find(s, \"barabara\") = " << String::find(s, "barabara") << endl;
	cout << "String::find_or_end(s, \"barabara\") = " << String::find_or_end(s, "barabara") << endl;
	cout << "String::find_last(s, \"barabara\") = " << String::find_last(s, "barabara") << endl;

	{
		cout << "String::replace(s, \"bar\", \" BAR \", 5) = ";
		const char * r = String::replace(s, "bar", " BAR ", 5);
		cout << r << endl;
		Memory::free(r);
	}
	{
		char t[String::len(s)];
		cout << "String::replace_inplace(t, \"bar\", \"...\", 5) = ";
		const char * r = String::replace_inplace(String::copy(t, s), "bar", "...", 5);
		cout << r << endl;
		assert(r == t);
	}
	{
		cout << "String::replace(s, \"barbar\", \" BARBAR \") = ";
		const char * r = String::replace(s, "barbar", " BARBAR ");
		cout << r << endl;
		Memory::free(r);
	}
	{
		char t[String::len(s)];
		cout << "String::replace_inplace(t, \"barbar\", \"...\") = ";
		const char * r = String::replace_inplace(String::copy(t, s), "barbar", "...");
		cout << r << endl;
		assert(r == t);
	}
	{
		cout << "String::replace(s, \"bax\", \" BAR \") = ";
		const char * r = String::replace(s, "bax", " BAR ");
		cout << r << endl;
		Memory::free(r);
	}
	{
		char t[String::len(s)];
		cout << "String::replace_inplace(t, \"bax\", \"...\") = ";
		const char * r = String::replace_inplace(String::copy(t, s), "bax", "...");
		cout << r << endl;
		assert(r == t);
	}
	{
		cout << "String::split(s, 'b' , 4):" << endl;
		for (auto p : String::split(s, 'b', 4)) {
			cout << " - " << p << endl;
			Memory::free(p);
		}
	}
	{
		char t[String::len(s)];
		cout << "String::split_inplace(t, 'b', 4):" << endl;
		for (auto p : String::split_inplace(String::copy(t, s), 'b', 4))
			cout << " - " << p << endl;
	}
	{
		cout << "String::split(s, \"b\" , 4):" << endl;
		for (auto p : String::split(s, "b", 4)) {
			cout << " - " << p << endl;
			Memory::free(p);
		}
	}
	{
		char t[String::len(s)];
		cout << "String::split_inplace(t, \"b\", 4):" << endl;
		for (auto p : String::split_inplace(String::copy(t, s), "b", 4))
			cout << " - " << p << endl;
	}
	{
		cout << "String::split(s, \"bar\"):" << endl;
		for (auto p : String::split(s, "bar")) {
			cout << " - " << p << endl;
			Memory::free(p);
		}
	}
	{
		char t[String::len(s)];
		cout << "String::split_inplace(t, \"bar\"):" << endl;
		for (auto p : String::split_inplace(String::copy(t, s), "bar"))
			cout << " - " << p << endl;
	}
	return 0;
}
