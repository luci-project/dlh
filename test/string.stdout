String::find(s, 'b') = barberbarbarabarbarbaren!
String::find_or_end(s, 'b') = barberbarbarabarbarbaren!
String::find_last(s, 'b') = baren!
String::find(s, "barbar") = barbarabarbarbaren!
String::find_or_end(s, "barbar") = barbarabarbarbaren!
String::find_last(s, "barbar") = barbaren!
String::find(s, 'y') = 
String::find_or_end(s, 'y') = 
String::find_last(s, 'y') = 
String::find(s, "barabara") = 
String::find_or_end(s, "barabara") = 
String::find_last(s, "barabara") = 
String::replace(s, "bar", " BAR ", 5) = Rha BAR ber BAR  BAR a BAR  BAR baren!
String::replace_inplace(t, "bar", "...", 5) = Rha...ber......a......baren!
String::replace(s, "barbar", " BARBAR ") = Rhabarber BARBAR a BARBAR baren!
String::replace_inplace(t, "barbar", "...") = Rhabarber...a...baren!
String::replace(s, "bax", " BAR ") = Rhabarberbarbarabarbarbaren!
String::replace_inplace(t, "bax", "...") = Rhabarberbarbarabarbarbaren!
String::split(s, 'b' , 4):
 - Rha
 - ar
 - er
 - ar
 - arabarbarbaren!
String::split_inplace(t, 'b', 4):
 - Rha
 - ar
 - er
 - ar
 - arabarbarbaren!
String::split(s, "b" , 4):
 - Rha
 - ar
 - er
 - ar
 - arabarbarbaren!
String::split_inplace(t, "b", 4):
 - Rha
 - ar
 - er
 - ar
 - arabarbarbaren!
String::split(s, "bar"):
 - Rha
 - ber
 - a
 - en!
String::split_inplace(t, "bar"):
 - Rha
 - ber
 - a
 - en!
